src/evaluators/source-4-1-lazy.js

```diff
@@ -0,0 +1,924 @@
+
+// The core of our evaluator is formed by the
+// implementation of function applications.
+// Applications are tagged with "application"
+// and have "operator" and "operands"
+
+function is_application(stmt) {
+   return is_tagged_list(stmt, "application");
+}
+function operator(stmt) {
+   return head(tail(stmt));
+}
+function operands(stmt) {
+   return head(tail(tail(stmt)));
+}
+function no_operands(ops) {
+   return is_null(ops);
+}
+function first_operand(ops) {
+   return head(ops);
+}
+function rest_operands(ops) {
+   return tail(ops);
+}
+      
+// primitive functions are tagged with "primitive"      
+// and come with a Source function "implementation"
+
+function make_primitive_function(impl) {
+    return list("primitive", impl);
+}
+function is_primitive_function(fun) {
+   return is_tagged_list(fun, "primitive");
+}
+function primitive_implementation(fun) {
+   return list_ref(fun, 1);
+}
+
+function actual_value(exp, env) {
+    //display("Callig actual_value");
+    //display(evaluate(exp, env));
+    return force_it(evaluate(exp, env));
+}
+
+function delay_it(exp, env) {
+    return list("thunk", exp, env);
+}
+
+function is_thunk(obj) {
+    return is_tagged_list(obj, "thunk");
+}
+
+function thunk_exp(thunk) {
+    return head(tail(thunk));
+}
+
+function thunk_env(thunk) {
+    return head(tail(tail(thunk)));
+}
+
+function is_evaluated_thunk(obj) {
+    return is_tagged_list(obj, "evaluated_thunk");
+}
+
+function thunk_value(obj) {
+    return head(tail(obj));
+}
+
+function force_it(obj) {
+    //display("Calling force it");
+    //display(obj);
+    if(is_thunk(obj)) {
+        
+        const result = actual_value(thunk_exp(obj), thunk_env(obj));
+        
+        set_head(obj, "evaluated_thunk");
+        set_head(tail(obj), result);
+        set_tail(tail(obj), null);
+        
+        return result;
+        
+    } else if(is_evaluated_thunk(obj)) {
+        
+        return thunk_value(obj);
+        
+    } else {
+        return obj;
+    }
+    
+}
+
+/* APPLY */
+
+// apply_in_underlying_javascript allows us
+// to make use of JavaScript's primitive functions
+// in order to access operators such as addition
+
+function apply_primitive_function(fun, argument_list) {
+    return apply_in_underlying_javascript(
+                primitive_implementation(fun),
+                argument_list);     
+}
+
+function list_of_arg_values(exps, env) {
+    return no_operands(exps)
+        ? null
+        : pair(actual_value(first_operand(exps), env),
+          list_of_arg_values(rest_operands(exps), env));
+}
+
+function list_of_delayed_args(exps, env) {
+    return no_operands(exps)
+        ? null
+        : pair(delay_it(first_operand(exps), env),
+          list_of_delayed_args(rest_operands(exps), env));
+}
+
+function is_pair_statement (fun){
+    return head(tail(operator(fun))) === "pair";
+}
+
+function is_head_statement(stmt)
+{
+    return head(tail(operator(stmt))) === "head";
+}
+
```

Comment:
Add a comment that refers to this textbook section: "Note that these lazy lists are even lazier than the streams of chapter 3: The head of the list, as well as the tail, is delayed.[2] In fact, even accessing the head or tail of a lazy pair need not force the value of a list element. The value will be forced only when it is really needed—e.g., for use as the argument of a primitive, or to be printed as an answer."
https://sicp.comp.nus.edu.sg/chapters/84#p4